man/GUI-tools/helptool.man
.so ../macros
.TH helptool GUI-tools %G% "BitMover, Inc." BitKeeper
.\"    ====================== BitKeeper helptool ======================
.SH NAME
bk helptool \- graphical front-end to the help system
.SH SYNOPSIS
.B bk helptool [
.IR topic ]
.SH DESCRIPTION
.LP
The helptool command is a graphical interface to the online help system.
.LP
There are two text windows displayed, a window of topics on the left
and the help text on the right.  There are multiple sections, which
group related topics together.  A topic is slightly indented, the
section heading is not.
.LP
You can move around using the mouse, clicking on topics.  You can
also page through the entire help using PageDown or PageUp.  To move
from section to section, use the Left arrow to move up and the Right
arrow to move forward.  This is fairly useful since the first entry
in each section tries to give a brief overview of that section.
.SH "HYPERLINK STACK"
.LP
Helptool has a stack of places that you have visited, very similar
to  the hyperlink stack in Netscape.  If you click on a topic you will
notice that the "Back" button becomes enabled.  Clicking it will move
back to the previous topic.  Similarly for forward.  The same hot keys
which work in Netscape are used here for forward and back.
.SH SEARCHING
.LP
Helptool includes a simple search facility.  To search for information,
you type a word or phrase and hit return (or click on Search).  The search
facility is fairly basic, single word searches tend to work better
than phrases.  If a phrase is wanted, it must be put in quotes, i.e.,
"lock file".
.LP
The lines which match, if any, are shown slightly indented.  Above the
indent lines is the name of the command or help page which contains
the lines.  You can click on that to jump to that page.  Each command
or page which contains the search phrase will be be highlighted in orange
so that you click through each of the pages which contain the information.
.LP
If you go to a page and it was not what you wanted, and you wish
to look at the summary lines again, jump back by clicking Search or
hitting return.
.SH BINDINGS
.LP
The up/down bindings will move through topics, so when they hit the end
and the key is pressed again, the topic changes to the next or previous.
.TP 15
.B DownArrow
Scroll the help text down one line.
.tp
.B UpArrow
Scroll the help text up one line.
.tp
.B PageDown
Scroll the help text down one screen.
.tp
.B PageUp
Scroll the help text up one screen.
.tp
.B Home
Scroll to the top of help text.
.tp
.B End
Scroll to the bottom of the help text.
.tp
.B Escape
Quit the helptool.
.tp
.B LeftArrow
Move to the beginning of the next section (not topic).
.tp
.B RightArrow
Move to the beginning of the previous section (not topic).
.tp
.B Alt-Left
Move backwards to the previously visited topic.
.tp
.B Alt-RightArrow
Move forwards to the to the next topic in the stack.
.tp
.B Control-e
Scroll the help text down one line.
.tp
.B Control-y
Scroll the help text up one line.
.SH BUGS
.LP
Unquoted phrases seem to just do the first word.
.SH "SEE ALSO"
.B bk help GUI


RELEASE-NOTES
============================================================================
Release notes for BitKeeper version 6.0.2 (released Oct 29, 2013)

    This is a bugfix release (one feature snuck in).

Bugfixes

  bk diffs
    - Fixed a bug where 'bk diffs -NL' did not list extra files.
    - bk diffs -s was not working because of the switch to -u as
      the default.
    - Function header matching in 'bk diffs -p' now works with C++
      names.

  bk difftool
    - When looking at diffs over different types of files, a binary
      file in the list would cause diffs for the next text file to not
      be displayed.

  bk gone
    - People would try and run 'bk gone foo.c' which is incorrect, the
      correct usage is:
	bk gone `bk log -r+ -nd:ROOTKEY:`
      bk gone now works harder to validate that the arguments passed
      are actually keys.

  Other bugfixes
    - Fix a bug in the BK6->BK5 compat code where some files were not
      converted correctly to the BK5 format.
    - Stale locks on windows were not always being cleaned immediately.
    - Ignore extra files in .bk which were causing problems in Mac OS.

Features

   bk annotate
    - MD5 keys are now supported in the annotation options by running
      either 'bk annotate -a5' or 'bk annotate -A5'.

============================================================================
Release notes for BitKeeper version 6.0.1 (released Aug 9, 2013)

    This is a bugfix release (a few features snuck in).

Bugfixes

  bk clone
    - 'bk clone -@' wasn't populating comps when the remote repository
      was a subset of the local baseline.
    - 'bk clone -r<rev> [--upgrade|--downgrade] repo localcopy' could
      fail in some cases, and has been fixed.
    - 'bk clone -r<rev> bk://host/repo localcopy' will now fail if the
      BK client version is bk-6.0 and the bkd is bk-6.0.1 or later.
      The solution is to upgrade the client.

  bk commit
    - 'bk commit -f' was still prompting for comments.
    - 'bk commit' had lost some status reporting; that has been
      restored.

  bk collapse
    - 'bk collapse' would leave components marked as pending; this caused
      pull to error with a "KEY MISMATCH" message.
    - 'bk fix/collapse' in a gate was only partially disabled, it would
      collapse in the product but would refuse to collapse in
      components.

  bk diffs
    - 'bk -Ucx diffs -N' would only print diffs for one new file;
      fixed to print diffs for all new files

  bk difftool
    - The file listing box on difftool resizes itself to be as large
      as the longest pathname shown (it used to truncate).
    - 'bk difftool -S -Lpath/to/component' is now supported.
      (-S == scan only the current component)
    - The 'next' button in bk difftool didn't come up highlighted when
      there was more than one file.
    - Better error reporting with -L when a single GCA doesn't exist.

  bk pull
    - A pull from a remote repository served by a bk version 5.x could fail
      if a component clone was part of the operation.  It works now.

  bk push
    - A push from a local repository served by a bk version 5.x to a remote
      repository that uses new storage format of bk-6 could fail if a component
      clone was part of the operation.  It works now.

  bk revtool
    - A particular sequence of clicks and launching difftool could
      make revtool unresponsive.

    - When launching revtool with a line number (bk revtoo +lno),
      the comments for the revision that added that line were not being shown.

  bk sfiles
    - In rare cases, a pending component would not be seen as pending
      ('bk -pA' did not list them but it does now).

  Misc
    - In rare cases, permissions in the .bk/SCCS directory could be set
      wrong and cause bk to fail.  The known case has been fixed and
      the problem is now repaired automatically.

    - In another permissions failure, when a user didn't have write
      permissions on the repository, commit would fail with an
      unhelpful error message.  The problem is now reported more
      clearly.

  Graphical tools
    - When resizing graphical tools quickly, the resize labels could
      get out of sync; they update more frequently now.

Features
    - bk difftool now supports -w/-b to ignore white space changes
      like 'bk diffs'.  There is toolbar control of this feature as
      well.

    - BitKeeper versions prior to 6.0 may now clone repositories in the new
      format; they will automatically be downgraded to a compatible format.

============================================================================
Release notes for BitKeeper version 6.0 (released Jul 05, 2013)

    The primary focus of the 6.0 effort is performance improvements.

New repository format for performance
    BK6 has a new file format designed for performance.

    The new repository format cannot be read by bk-5.x releases
    of BitKeeper, but clone can convert a new repository to a bk-5.x
    readable repository.

Compatibility
    New file formats are a cause for concern but we've done the extra
    work so you may try it without forcing everyone to upgrade. BK6 can
    pull/push the new format from/to bk5's format.

    To enable the the new, faster format:
  
	bk clone --upgrade bk5-repo bk6-repo
  
    To downgrade to be compatible with bk5 (or bk4 if it is a non-nested
    repo):

	bk clone --downgrade bk6-repo bk5-repo

    To allow clones/pulls/pushes from bk4 & bk5, serve the repositories
    with a bk6 bkd:

    	cd bk6-repo
	bk pull bk://bk6-server/bk5-repo
	bk push bk://bk6-server/bk5-repo
 
Performance improvements:
    In the examples below a non-nested test repository contained 55,000
    files with 282,000 changesets on an NFS server.  Both the server
    and client caches were dropped (made cold) before each operation.

    Note that none of the performance improvements depend on the use of
    a file system notifier.  We considered and rejected such an approach

src/gui/helptool.tcl
# helptool - a tool for showing BK help
# Copyright (c) 1999 by Larry McVoy; All rights reserved

proc main {} \
{

	bk_init
	loadState help
	widgets
	restoreGeometry "help" 
	getHelp

	# This must be done after getFiles, because getFiles may cause the
	# app to exit. If that happens, the window size will be small, and
	# that small size will be saved. We don't want that to happen. So,
	# we only want this binding to happen if the app successfully starts
	# up
	bind . <Destroy> {
		if {[string match %W "."]} {
			saveState help
		}
	}

	after idle [list wm deiconify .]
	after idle [list focus -force .]
}

#
# Sets the global 'line' to the x'th line after 'line'
#
proc doNext {x} \
{
	global	line nTopics

	if {($x == -1) && ($line == 2.0)} { return }
	if {($x == 1) && ($line == 0.0)} {
		set l 1.0
	} else {
		set l [.ctrl.topics index "$line + $x lines"]
	}
	if {$l > $nTopics} { return }
	set line $l
	doSelect $x
}

proc doNextSection {x} \
{
	global	line nTopics

	if {($x == -1) && ($line == 2.0)} { return }
	if {($x == 1) && ($line == 0.0)} {
		set line 1.0
		doSelect $x
		return
	} 
	set l [.ctrl.topics index "$line + $x lines"]
	if {$l > $nTopics} { return }
	set topic [.ctrl.topics get $l "$l lineend"]
	while {[regexp {^ } $topic]} {
		set l [.ctrl.topics index "$l + $x lines"]
		if {$l > $nTopics} { return }
		set topic [.ctrl.topics get $l "$l lineend"]
	}
	set line $l

	# If going backwards, go back to the first line.
	if {$x == -1 && $line != 1.0} {
		set line [.ctrl.topics index "$line + $x lines"]
		set topic [.ctrl.topics get $line "$line lineend"]
		while {[regexp {^ } $topic]} {
			set line [.ctrl.topics index "$line + $x lines"]
			set topic [.ctrl.topics get $line "$line lineend"]
		}
		.ctrl.topics see $line
	}
	doNext 1
}

proc doPixSelect {x y} \
{
	global	line stackMax stackPos

	set stackMax $stackPos
	set line [.ctrl.topics index "@$x,$y linestart"]
	doSelect 1
}

#
# Selects the x'th line, tags it as selected, calls bkhelp for that topic
#
proc doSelect {x {search {}}} \
{
	global	line stack stackMax stackPos

	busy 1
	.ctrl.topics see $line
	set topic [.ctrl.topics get $line "$line lineend"]
	if {$topic == ""} { busy 0; return }
	if {[regexp {^ } $topic] == 0} {
		doNext $x
		return
	}
	.ctrl.topics tag remove "select" 1.0 end
	.ctrl.topics tag add "select" $line "$line lineend + 1 char"
	.ctrl.topics tag raise "select"
	bkhelp $topic $search
	# Don't increment if we are going to where we are
	if {$stackPos == 0 || $stack($stackPos) != $line} { incr stackPos }
	set stack($stackPos) $line
	if {$stackMax < $stackPos} { set stackMax $stackPos }
	if {$stackPos > 1} {
		.menu.back configure -state normal
	} else {
		.menu.back configure -state disabled
	}
	if {$stackMax > $stackPos} {
		.menu.forw configure -state normal
	} else {
		.menu.forw configure -state disabled
	}
	busy 0
}

# set the top of stack to where we are now, we are heading in a new direction.
proc stackReset {} \
{
	global	stackMax stackPos

	set stackMax $stackPos
}

# Pop up the stack one if we can.
proc upStack {} \
{
	global	line stack stackMax stackPos

	if {$stackPos > 1} {
		incr stackPos -1
		set line $stack($stackPos)
		# because doSelect will push again
		incr stackPos -1
		doSelect 1
	}
}

proc downStack {} \
{
	global	line stack stackMax stackPos

	if {$stackMax > $stackPos} {
		incr stackPos
		set line $stack($stackPos)
		# because doSelect will push again
		incr stackPos -1
		doSelect 1
	}
}

proc bkhelp {topic {search {}}} \
{
	global line line2full gc

	set msg "BitKeeper help -- $topic"
	wm title . $msg
	set f [open "| bk help $gc(help.helptext) -p $topic"]
	.text.help configure -state normal
	.text.help delete 1.0 end
	set lineno 1
	set yview ""
	while {[gets $f help] >= 0} {
		.text.help insert end "$help\n"
		if {[regexp {^[A-Z][ \t\nA-Z.!?|()-]+$} $help]} {
			set i "$lineno.0"
			.text.help tag add "bold" $i "$i lineend"
		}
		if {$search != "" && [regexp $search $help]} {
			set search ""
			set yview "$lineno.0"
		}
		incr lineno
	}
	set i "$lineno.0"
	.text.help delete "$i" end
	catch {close $f} dummy
	.text.help configure -state disabled
	bk_highlight
	if {$yview != ""} { .text.help yview $yview }
}

proc highlight {tag index len} \
{
	set index2 [.text.help index "$index + $len chars"]
	.text.help tag add $tag $index $index2
	.text.help tag add seealso $index $index2
	.text.help tag bind \
	    $tag <Button-1> "getSelection $tag; stackReset; doSelect 1"
}

# Look for each <bk> word, then find the next word,
# if next_word == help|helptool
#	if word is a topic
#		highlight that word
#	else
#		highlight help|helptool
# else
#	if word is a topic
#		highlight that word
#	else
#		highlight "bk"
# XXX - maybe recode this in C as an option to "bk help"?
proc bk_highlight {} \
{
	set index 4.0
	set t .text.help
	while {"$index" != ""} {
		set index [$t search \
		    -count bklen -regexp {(^| |`|"|/)bk([ ]+|$)} $index end]
		if {"$index" == ""} { break }

		# Get start of "bk"
		if {[$t get $index] == " "} {
			set bkindex [$t index "$index + 1 chars"]
		} else {
			set bkindex $index
		}

		# skip past " bk ".
		set index [$t index "$index + $bklen chars"]

		# Get next word
		set w1index [$t search \
		    -count w1len -regexp {[a-zA-Z0-9_\-]+} $index end]
		if {"$w1index" == ""} {
			highlight bk $bkindex 2
			break
		}
		set w1 [$t get $w1index [$t index "$w1index + $w1len chars"]]
#puts "WORD1=$w1 @ $w1index .. $w1len"
		if {[regexp {^(help|helptool)$} $w1]} {
			# skip past "$w1".
			set index [$t index "$w1index + $w1len chars"]

			# Get next word
			set w2index [$t search \
			    -count w2len -regexp {[a-zA-Z0-9_\-]+} $index end]
			if {"$w2index" == ""} {
				highlight $w1 $w1index $w1len
				break
			}
			set w2 [$t get \
			    $w2index [$t index "$w2index + $w2len chars"]]
#puts "WORD2=$w2 @ $w2index .. $w2len"
			if {[topic2line $w2] != ""} {
				highlight $w2 $w2index $w2len
				set index [$t index "$w2index + $w2len chars"]
			} else {
				highlight $w1 $w1index $w1len
				set index [$t index "$w1index + $w1len chars"]
			}
			continue
		}

		if {[topic2line $w1] != ""} {
			highlight $w1 $w1index $w1len
			set index [$t index "$w1index + $w1len chars"]
		} else {
			highlight bk $bkindex 2
			set index [$t index "$bkindex + 2 chars"]
		}
	}
}

proc topic2line {key} \
{
	global lines aliases full

	set l ""
	catch { set l $lines($key) } dummy
	if {"$l" == ""} { catch { set l $lines($aliases($key)) } dummy }
	if {"$l" != ""} { set l "$l.0" }
	return $l
}

proc search {} \
{
	global	search_word lines gc opts

	if {$search_word == "" } {
		set search_word \
		    "REPLACE all of this with a word, and click search"
		update
		return
	}
	if ($gc(help.exact)) {
		set opts ""
	} else {
		set opts " -a"
	}
	.ctrl.topics tag remove "select" 1.0 end
	.ctrl.topics tag remove "search" 1.0 end
	.text.help configure -state normal
	.text.help delete 1.0 end
	set f [open "| bk helpsearch $gc(help.helptext)$opts -l $search_word" "r"]
	set last ""
	while {[gets $f line] >= 0} {
		set tab [string first "\t" $line"]
		set key [string range $line 0 [expr {$tab - 1}]]
		incr tab
		set sentence [string range $line $tab end]
		set sentence [string trim $sentence]
		if {$last != $key} {
			set last $key
			.text.help insert end "$key\n" "$key seealso"
			.text.help tag bind $key <Button-1> \
			    "getSelection $key; stackReset; doSelect 1; break"
			set l [topic2line $key]
			if {"$l" != ""} {
				.ctrl.topics tag add "search" \
				    "$l linestart" "$l lineend + 1 char"
			}
		}
		.text.help insert end "  $sentence\n"
	}
	catch {close $f} dummy
	.text.help configure -state disabled
	.text.help tag configure seealso -foreground $gc(help.linkColor) \
	    -underline true
}

proc clearSearch {} \
{
	global search_word

	set search_word ""
	.ctrl.topics tag remove "search" 1.0 end
	.text.help configure -state normal
	.text.help delete 1.0 end
	.text.help configure -state disabled
	doSelect 1
}

proc scroll {what dir} \
{
	global	gc line

	set a [lindex [.text.help yview] 0]
	set b [lindex [.text.help yview] 1]
	if {$dir == 1 && $b == 1} {
		doNext 1
	} elseif {$dir == -1 && $a == 0 && $line > 1.0} {
		doNext -1
	} elseif {$what == "page"} {
		set x [expr $gc(help.height) - 1]
		set x [expr $x * $dir]
		.text.help yview scroll $x units
	} else {
		.text.help yview scroll $dir units
	}
}

proc widgets {} \
{
	global	line gc firstConfig pixelsPerLine
	global	search_word stackMax stackPos d

	set stackMax 0
	set stackPos 0

	getConfig "help"
	if {$gc(windows) || $gc(aqua)} {
		set py 0
	} else {
		set py 1
	}

	set rootX [winfo screenwidth .]
	set rootY [winfo screenheight .]
	wm title . "BitKeeper Help"
	set firstConfig 1
	set sb $gc(help.scrollbars)

	ttk::frame .menu
	    ttk::button .menu.done -text "Quit" -command { exit }
	    ttk::button .menu.help -text "Help" -command {
		global	line

		clearSearch
		set line [topic2line helptool]
		doSelect 1
	    }
	    ttk::button .menu.back -text "Back" -state disabled \
		-command { upStack }
	    ttk::button .menu.forw -text "Forward" -state disabled \
		-command { downStack }
	    ttk::button .menu.clear -text "Clear Search" \
		-command { clearSearch }
	    ttk::button .menu.search -text "Search:" -command { search }
	    ttk::entry .menu.entry -textvariable search_word
	    grid .menu.done -row 0 -column 0 -sticky ew -padx 1
	    grid .menu.help -row 0 -column 1 -sticky ew -padx 1
	    grid .menu.back -row 0 -column 3 -sticky ew -padx 1
	    grid .menu.forw -row 0 -column 4 -sticky ew -padx 1
	    grid .menu.clear -row 0 -column 5 -sticky ew -padx 1
	    grid .menu.search -row 0 -column 6 -sticky ew -padx 1
	    grid .menu.entry -row 0 -column 7 -sticky ew -padx 1
	    grid columnconfigure .menu 7 -weight 1
	ttk::frame .ctrl
	    text .ctrl.topics \
	        -borderwidth 1 \
		-spacing1 1 \
		-spacing3 1 \
		-wrap none \
		-height $gc(help.height) \
		-font $gc(help.fixedFont) -width 16 \
		-background "#F8F8F8" \
		-yscrollcommand [list .ctrl.yscroll set] \
		-xscrollcommand [list .ctrl.xscroll set]
	    ttk::scrollbar .ctrl.yscroll -orient vertical \
		-command ".ctrl.topics yview"
	    ttk::scrollbar .ctrl.xscroll -orient horizontal \
		-command ".ctrl.topics xview"

	    if {[string index $sb 0] eq "R"} {
		    grid .ctrl.topics -row 0 -column 0 -sticky nsew
		    grid .ctrl.yscroll -row 0 -column 1 -sticky nse
		    grid .ctrl.xscroll -row 1 -column 0 -sticky ew
	    } else {
		    grid .ctrl.topics -row 0 -column 1 -sticky nsew
		    grid .ctrl.yscroll -row 0 -column 0 -sticky nse
		    grid .ctrl.xscroll -row 1 -column 1 -sticky ew
	    }
	    grid rowconfigure    .ctrl .ctrl.topics -weight 1
	    grid columnconfigure .ctrl .ctrl.topics -weight 1

	ttk::frame .text
	    text .text.help \
	        -borderwidth 1 \
		-wrap none \
		-font $gc(help.fixedFont) \
		-width $gc(help.width) \
		-height $gc(help.height) \
		-padx 4 \
		-background $gc(help.textBG) -fg $gc(help.textFG) \
		-xscrollcommand [list .text.x2scroll set] \
		-yscrollcommand [list .text.y2scroll set]

	    ttk::scrollbar .text.x2scroll -orient horizontal \
		-command ".text.help xview"
	    ttk::scrollbar .text.y2scroll -orient vertical \
		-command ".text.help yview"

	    if {[string index $sb 1] eq "R"} {
		    grid .text.help     -row 0 -column 0 -sticky nsew
		    grid .text.y2scroll -row 0 -column 1 -sticky nse
		    grid .text.x2scroll -row 1 -column 0 -sticky ew 
	    } else {
		    grid .text.help     -row 0 -column 1 -sticky nsew
		    grid .text.y2scroll -row 0 -column 0 -sticky nse
		    grid .text.x2scroll -row 1 -column 1 -sticky ew 
	    }

	    grid rowconfigure    .text .text.help -weight 1
	    grid columnconfigure .text .text.help -weight 1

	grid .menu -row 0 -column 0 -columnspan 2 -sticky nsew -pady 2
	grid .ctrl -row 1 -column 0 -sticky nsew
	grid .text -row 1 -column 1 -sticky nsew

	grid rowconfigure . 1 -weight 1
	grid columnconfigure . 0 -weight 0
	grid columnconfigure . 1 -weight 1

	bind .ctrl.topics <Button-1>	{ doPixSelect %x %y; break }
	bind .ctrl.topics <Button-2>	{ doPixSelect %x %y; break }
	bind .ctrl.topics <Button-3>	{ doPixSelect %x %y; break }
	bind .ctrl.topics <Motion>	"break"
	bind .text.help <ButtonPress>	"focus .text.help"
	bind all <Control-e>		{ scroll "line" 1; break }
	bind all <Control-y>		{ scroll "line" -1; break }
	bind all <Control-b>		{ scroll "page" -1 ; break}
	bind all <Control-f>		{ scroll "page"  1 ; break}
	bind all <Down>			{ scroll "line" 1; break }
	bind all <Up>			{ scroll "line" -1; break }
	bind .text.help <Down>		{ scroll "line" 1; break }
	bind .text.help <Up>		{ scroll "line" -1; break }
	bind all <Left>			".text.help xview scroll -1 units;break"
	bind all <Right>		".text.help xview scroll 1 units; break"
	bind all <Prior>		{ scroll "page" -1; break }
	bind all <Next>			{ scroll "page" 1; break }
	bind Text <Prior>		{ scroll "page" -1; break }
	bind Text <Next>		{ scroll "page" 1; break }
	bind all <Home>		 	".text.help yview -pickplace 1.0; break"
	bind all <End>		 	".text.help yview -pickplace end; break"
	bind all <Control-Up>		{ doNext -1 }
	bind all <Control-Down>		{ doNext 1 }
	bind all <Control-Left>	 	"doNextSection -1"
	bind all <Control-Right> 	"doNextSection 1"
	bind all <Alt-Left>		{ upStack }
	bind all <Alt-Right>		{ downStack }
	bind all <$gc(help.quit)>	{ exit }
	if {$gc(aqua)} {
		bind all <Command-q> exit
		bind all <Command-w> exit
	}
	bind .menu.entry <Return> { search }
	bindtags .menu.entry { all .menu.entry TEntry . }
 	bindtags .ctrl.topics {.ctrl.topics . all}
	#bindtags .text.help {.text.help . all}
	bind .text.help <Configure> {
		global	gc pixelsPerLine firstConfig

		set x [winfo height .text.help]
		# This gets executed once, when we know how big the text is
		if {$firstConfig == 1} {
			set h [winfo reqheight .text.help]
			set pixelsPerLine [expr {$h / $gc(help.height)}]
			set firstConfig 0
		}
		set x [expr {$x / $pixelsPerLine}]
		set gc(help.height) $x
	}
	.ctrl.topics tag configure "select" -background $gc(help.selectColor) \
	    -relief ridge -borderwid 1
	.text.help tag configure "title" -background #8080c0
	.text.help tag configure "bold" -font $gc(help.fixedBoldFont)
	.text.help tag configure seealso -foreground $gc(help.linkColor) \
	    -underline true
	.ctrl.topics tag configure "search" -background $gc(help.topicsColor) \
	    -relief ridge -borderwid 1

	after idle [list after 1 focus .menu.entry]
}

proc busy {busy} \
{
	if {$busy != 0} {
		. configure -cursor watch
		.text.help configure -cursor watch
		.ctrl.topics configure -cursor watch
	} else {
		. configure -cursor left_ptr
		.text.help configure -cursor left_ptr
		.ctrl.topics configure -cursor left_ptr
	}
	update
}

proc getSelection {page} \
{
	global line lines aliases

	set l ""
	catch { set l $lines($page) } dummy
	if {"$l" == ""} { catch { set l $lines($aliases($page)) } dummy }
	if {"$l" == ""} {
		puts "No help for $page"
		exit
	}
	incr l -1
	set line [.ctrl.topics index "1.0 + $l lines"]
}

proc getHelp {} \
{
	global	nTopics argv line lines aliases line2full gc

	set nTopics 0
	set file [lindex $argv 0]
	if {[regexp {^-f} $file]} {
		if {$file == "-f"} {
			set file [lindex $argv 1]
			set gc(help.helptext) "-f$file"
			set keyword [lindex $argv 2]
			set search [lindex $argv 3]
		} else {
			set gc(help.helptext) $file
			set keyword [lindex $argv 1]
			set search [lindex $argv 2]
		}
	} else {
		set keyword [lindex $argv 0]
		set search [lindex $argv 1]
	}
	set f [open "| bk helptopics $gc(help.helptext)"]
	.ctrl.topics configure -state normal
	set section ""
	while {[gets $f topic] >= 0} {
		if {$topic == "Aliases"} { break }
		.ctrl.topics insert end "$topic\n"
		if {[string index $topic 0] != " "} {
			set section $topic
			set topic ""
		} else {
			set topic [string trim $topic]
		}
		incr nTopics 1
		set lines($topic) $nTopics
		if {$topic != ""} {
			set full "$section/$topic"
		} else {
			set full $section
		}
		set lines($full) $nTopics
		set index "$nTopics.0"
		set line2full($index) $full
	}
	while {[gets $f topic] >= 0} {
		set l [split $topic \t]
		set key [lindex $l 0]
		set val [lindex $l 1]
		set aliases($key) $val
		# Store the short key as well.  This can cause problems
		# if there are name space collisions, but then use full names.
		set slash [string first "/" $key]
		incr slash
		set short [string range $key $slash end]
		set aliases($short) $val
	}
	catch {close $f} dummy
	.ctrl.topics configure -state disabled
	.text.help configure -state disabled
	if {$keyword == ""} {
		getSelection Common
	} else {
		getSelection $keyword 
	}
	doSelect 1 $search
	if {$keyword == ""} {
		.ctrl.topics yview moveto 0
	}
}

main
